lib/src/grinder.dart

```diff
@@ -0,0 +1,224 @@
+   *
+   * [start] should be called once and only once; i.e., Grinder instances are
+   * not re-usable.
+   *
+   * Items in [invocations] can either be [String] names of tasks to invoke, or
+   * full [TaskInvocation]s.
+   *
+   * The [dontRun] parameter can be used to audit the grinder file, without
+   * actually executing any targets.
+   *
+   * Throws [GrinderException] if named tasks don't exist, or there are
+   * cycles in the dependency graph.
+   */
+  Future start(Iterable invocations, {bool dontRun: false}) {
+    if (!dontRun && _taskDeps != null) {
+      throw new StateError("Grinder instances are not re-usable");
+    }
+
+    invocations = invocations.map((invocation) =>
+        invocation is String ? new TaskInvocation(invocation) : invocation);
+
+    DateTime startTime = new DateTime.now();
+
+    if (invocations.isEmpty) {
+      if (defaultTask != null) {
+        invocations = [new TaskInvocation(defaultTask.name)];
+      } else if (!dontRun) {
+        throw new GrinderException('Tried to call non-existent default task.');
+      }
+      if (invocations.isEmpty) return new Future.value();
+    }
+
+    // Verify that all named tasks exist.
+    for (var invocation in invocations) {
+      var name = invocation.name;
+      if (getTask(name) == null) {
+        throw new GrinderException("task '$name' doesn't exist");
+      }
+    }
+
+    // Verify that there aren't any duplicate names.
+    Set<String> names = new Set();
+
+    for (GrinderTask task in _tasks) {
+      if (names.contains(task.name)) {
+        throw new GrinderException("task '${task.name}' is defined twice");
+      }
+      names.add(task.name);
+    }
+
+    // Verify that all referenced tasks exist.
+    for (GrinderTask task in tasks) {
+      for (var invocation in task.depends) {
+        if (getTask(invocation.name) == null) {
+          throw new GrinderException(
+              "task '${invocation.name}' referenced by ${task}, doesn't exist");
+        }
+      }
+    }
+
+    _calculateAllDeps();
+
+    // Verify that there are no dependency cycles.
+    for (GrinderTask task in tasks) {
+      if (getAllDependencies(task)
+      .any((invocation) => invocation.name == task.name)) {
+        throw new GrinderException("Task ${task} has a dependency cycle.\n"
+        "  ${task} ==> ${getAllDependencies(task).join(', ')}");
+      }
+    }
+
+    invocations.forEach(_postOrder);
+
+    if (!dontRun) {
+      log('grinder running ${_invocationOrder.join(' ')}');
+      log('');
+
+      return Future.forEach(_invocationOrder, (task) {
+        return _invokeTask(task);
+      }).then((_) {
+        Duration elapsed = new DateTime.now().difference(startTime);
+        log('finished in ${elapsed.inMilliseconds / 1000.0} seconds.');
+      });
+    } else {
+      return new Future.value();
+    }
+  }
+
+  /// Given a task, return all of its immediate dependencies.
+  Iterable<TaskInvocation> getImmediateDependencies(GrinderTask task) =>
```

Comment:
Cool, thanks for the explanation. I think my concern here was that we're mixing tasks and task invocations. So we're querying if Task a depends on TaskInvocation b, instead of Task a depending on Task b or TaskInvocation a depending on TaskInvocation b.